OrbitVulkanLayer/TimerQueryPool.h

```diff
@@ -0,0 +1,157 @@
+  [[nodiscard]] bool NextReadyQuerySlot(VkDevice device, uint32_t* allocated_index) {
+    absl::WriterMutexLock lock(&mutex_);
+    CHECK(device_to_potential_next_free_index_.contains(device));
+    CHECK(device_to_query_slots_.contains(device));
+    uint32_t potential_next_free_slot = device_to_potential_next_free_index_.at(device);
+    std::vector<SlotState>& slots = device_to_query_slots_.at(device);
+    uint32_t current_slot = potential_next_free_slot;
+    do {
+      if (slots.at(current_slot) == SlotState::kReadyForQueryIssue) {
+        device_to_potential_next_free_index_[device] = (current_slot + 1) % num_timer_query_slots_;
+        slots.at(current_slot) = SlotState::kQueryPendingOnGpu;
+        *allocated_index = current_slot;
+        return true;
+      }
+      current_slot = (current_slot + 1) % num_timer_query_slots_;
+    } while (current_slot != potential_next_free_slot);
+
+    return false;
+  }
+
+  // Resets an occupied slot to be ready for queries again. It will also call to Vulkan to reset the
+  // content of that slot (in contrast to `RollbackPendingQuerySlots`).
+  //
+  // Note that the pool must be initialized using `InitializeTimerQueryPool` before.
+  // Further, the given slots must be in the `kReadyForQueryIssue` state, i.e. must be a result
+  // of `NextReadyQuerySlot` and must not have been reset yet.
+  void ResetQuerySlots(VkDevice device, const std::vector<uint32_t>& slot_indices) {
+    ResetQuerySlots(device, slot_indices, false);
+  }
+
+  // Resets an occupied slot to be ready for queries again. It will *not* call to Vulkan to reset
+  // the content of that slot (in contrast to `ResetQuerySlots`).
+  // This is useful, if the slot was retrieved, but the actual query was not yet submitted to
+  // Vulkan (e.g. if on resetting the command buffer).
+  //
+  // Note that the pool must be initialized using `InitializeTimerQueryPool` before.
+  // Further, the given slots must be in the `kReadyForQueryIssue` state, i.e. must be a result
+  // of `NextReadyQuerySlot` and must not have been reset yet.
+  void RollbackPendingQuerySlots(VkDevice device, const std::vector<uint32_t>& slot_indices) {
+    ResetQuerySlots(device, slot_indices, true);
+  }
+
+ private:
+  enum class SlotState { kReadyForQueryIssue = 0, kQueryPendingOnGpu };
+
+  // Resets an occupied slot to be ready for queries again.
+  // If `rollback_only` is set, it will not call to Vulkan to reset the content of that slot.
+  // This is useful, if the slot was retrieved, but the actual query was not yet submitted to
+  // Vulkan (e.g. if on resetting the command buffer).
+  //
+  // Note that the pool must be initialized using `InitializeTimerQueryPool` before.
+  // Further, the given slots must be in the `kReadyForQueryIssue` state, i.e. must be a result
+  // of `NextReadyQuerySlot` and must not have been reset yet.
+  void ResetQuerySlots(VkDevice device, const std::vector<uint32_t>& slot_indices,
```

Comment:
I think I would avoid the overloading and call this something like `ResetQuerySlotsInternal`.

Reply:
Done.